tools/tsTreeshaker.ts

```diff
@@ -0,0 +1,419 @@
+    if (this.outDir === undefined) throw new Error('outDir is undefined');
+    const relativePath = path.relative(this.baseUrl, filePath);
+    return path.join(this.outDir, relativePath);
+  }
+  setCompilerOptions(options: ts.CompilerOptions) {
+    if (this.graph) {
+      if (this.baseUrl || this.outDir) return;
+    }
+
+    this.baseUrl ??= options.baseUrl || __dirname;
+    this.outDir ??= options.outDir || this.baseUrl;
+
+    this.sourceRoots = this.sourceRoots
+      .map((sourceRoot) => this.normalizePath(this.convertToDestinationPath(sourceRoot)));
+
+    const declarationRoots = [...this.sourceRoots]
+      .map((sourceRoot) => this.replaceExtension(sourceRoot, '.d.ts'));
+
+    const jsRoots = [...this.sourceRoots]
+      .map((sourceRoot) => this.replaceExtension(sourceRoot, '.js'));
+
+    this.graph = new DependencyGraph([...declarationRoots, ...jsRoots]);
+  }
+  setForceIncludeFiles(...files: string[]) {
+    if (this.graph === undefined) throw new Error('graph is undefined');
+
+    const projectPath = process.cwd();
+    files = files.map((file) =>
+      this.normalizePath(
+        this.convertToDestinationPath(path.resolve(projectPath, file))
+      )
+    );
+
+    this.forceIncludeFiles = [
+      ...files.map((file) => this.replaceExtension(file, '.d.ts')),
+      ...files.map((file) => this.replaceExtension(file, '.js'))
+    ];
+  }
+  addNode(filePath: string, requestedModule: string, extension: '.d.ts' | '.js') {
+    if (this.graph === undefined) throw new Error('graph is undefined');
+
+    requestedModule = this.getAbsolutePath(filePath, requestedModule);
+
+    try {
+      if (fs.lstatSync(requestedModule).isDirectory()) {
+        requestedModule = path.join(requestedModule, 'index');
+      }
+    } catch (e) {
+      // ignore
+    }
+
+    requestedModule = requestedModule + this.getFileExtension(requestedModule);
+    requestedModule = this.replaceExtension(requestedModule, extension);
+
+    filePath = this.replaceExtension(filePath, extension);
+
+    requestedModule = this.convertToDestinationPath(requestedModule);
+    requestedModule = this.normalizePath(requestedModule);
+    filePath = this.convertToDestinationPath(filePath);
+    filePath = this.normalizePath(filePath);
+
+    this.graph.addDependency(filePath, requestedModule);
+  }
+  getIncludedDependencies() {
+    if (this.graph === undefined) throw new Error('graph is undefined');
+
+    if (this.forceIncludeFiles !== undefined) {
+      return [
+        ...this.graph.getIncludedDependencies(),
+        ...this.forceIncludeFiles
+      ];
+    }
+    return this.graph.getIncludedDependencies();
+  }
+  isIncludedDependency(filePath: string, basePath?: string, overrideExtension?: string) {
+    if (this.graph === undefined) throw new Error('graph is undefined');
+
+    if (basePath !== undefined) {
+      filePath = this.getAbsolutePath(basePath, filePath);
+    }
+
+    try {
+      if (fs.lstatSync(filePath).isDirectory()) {
```

Comment:
Why catch error here?

Reply:
Because it throws error if directory does not exist

And if there is no directory, we should check if it is a file that including extension.
The existence of the file is guaranteed by the TypeScript compiler. If the import is incomplete in the first place, it can't be compiled

Reply:
Can it be converted into check but not `try...catch`?

`if (fs.exist(filePath) && fs.lstatSync(filePath).isDirectory())`

Reply:
okay